[REVIEW] host_memory_resource

jrhemstad opened the pull request:
Towards partial completion of https://github.com/rapidsai/rmm/issues/260, this PR introduces `host_memory_resource` to control host memory management.

- [x] Decide if `host_memory_resource` should have an `alignment` parameter for consistency with `std::pmr::memory_resource`
   - Resolved that it should.
- [x] Implementation
- [x] Documentation
- [x] Add pinned memory resource
- [x] Add new_delete memory resource
- [x] Tests
- [x] Add `aligned.hpp` for aligned memory allocation/free

harrism commented:
Discussion of this here: https://github.com/rapidsai/cudf/issues/4020 Suggest @vuule or @OlivierNV review this PR for suitability.

harrism commented on include/rmm/mr/host/host_memory_resource.hpp:
```diff
@@ -0,0 +1,168 @@
+ * When C++17 is available for use in RMM, `rmm::host_memory_resource` should
+ * inherit from `std::pmr::memory_resource`.
+ *
+ * This class serves as the interface that all host device memory resource
```
host device?

harrism commented on include/rmm/mr/host/host_memory_resource.hpp:
```diff
@@ -0,0 +1,168 @@
+   *`p`.
+   * @param stream Stream on which to perform deallocation
+   *---------------------------------------------------------------------------**/
+  void deallocate(void* p, std::size_t bytes,
```
Should this function be `noexcept`?

jrhemstad replied:
I'd thought so too, but for whatever reason neither [`std::pmr::memory_resource::deallocate`](https://en.cppreference.com/w/cpp/memory/memory_resource/deallocate), [`do_deallocate`](https://en.cppreference.com/w/cpp/memory/memory_resource/do_deallocate), nor the concrete derived types mark their [`deallocate`](https://en.cppreference.com/w/cpp/memory/synchronized_pool_resource/do_deallocate) functions with `noexcept`. 

I'm not really sure why.

jrhemstad replied:
I learned the answer is due to the ["Lakos Rule" ](https://quuxplusone.github.io/blog/2018/04/25/the-lakos-rule/)

> Remove noexcept specifications from each library function having a narrow contract, typically (but not always) indicated by the presence of a Requirements: clause.

Basically, since `deallocate` has undefined behavior if anything other than a pointer returned from `allocate` is passed in (a requirement), that means it shouldn't be marked `noexcept`.

harrism commented on include/rmm/mr/host/new_delete_resource.hpp:
```diff
@@ -0,0 +1,109 @@
+    // plus store of the correction offset
+    void *p = ::operator new(bytes + alignment + sizeof(std::size_t));
+    std::size_t ptr_int = reinterpret_cast<std::size_t>(p);
+    // calculate the offset, i.e. how many bytes of correction was necessary
```
```suggestion
    // calculate the offset, i.e. how many bytes of correction were necessary
```

harrism replied:
Or "are".

harrism commented on include/rmm/mr/host/pinned_memory_resource.hpp:
```diff
@@ -0,0 +1,110 @@
+      throw std::bad_alloc{};
+    }
+
+    std::size_t ptr_int = reinterpret_cast<std::size_t>(p);
```
This logic is used in two different memory resource implementations. Perhaps factor it out for reuse?

jrhemstad replied:
Yeah, I was just starting to work on that now :)

jrhemstad replied:
Done. See `aligned_allocate/deallocate`.

raydouglass commented:
rerun tests

trevorsm7 commented on include/rmm/mr/host/pinned_memory_resource.hpp:
```diff
@@ -0,0 +1,110 @@
+    // calculate the return pointer
+    char *ptr = static_cast<char *>(p) + offset;
+    // store the offset right after the actually returned value
+    std::size_t *offset_store = reinterpret_cast<std::size_t *>(ptr + bytes);
```
```suggestion
    *reinterpret_cast<std::size_t *>(ptr + bytes) = offset;
```

trevorsm7 commented on include/rmm/mr/host/pinned_memory_resource.hpp:
```diff
@@ -0,0 +1,110 @@
+    }
+    char *ptr = static_cast<char *>(p);
+    // calculate where the offset is stored
+    std::size_t *offset = reinterpret_cast<std::size_t *>(ptr + bytes);
```
```suggestion
    auto const offset = *reinterpret_cast<std::size_t *>(ptr + bytes);
```

trevorsm7 commented on tests/mr/host/mr_tests.cpp:
```diff
@@ -0,0 +1,239 @@
+  EXPECT_NO_THROW(this->mr->deallocate(p, 0));
+}
+
+TYPED_TEST(MRTest, AllocateWord) {
```
Looks like AllocateWord, AllocateKB, AllocateMB, and AllocateGB could be refactored to a single function that takes size as a param

harrism replied:
Good point. Could probably also re-use a bunch of existing device_memory_resource tests by parameterizing on memory_resource type?

jrhemstad replied:
`device_memory_resource` and `host_memory_resource` can't really shared the same tests because `host_memory_resource` allocation uses an `alignment` parameter whereas `device_memory_resource` alignment is fixed to 256. Likewise, the device memory resource tests test for if the memory is device memory, which will obviously fail for host memory resources.

harrism replied:
> Looks like AllocateWord, AllocateKB, AllocateMB, and AllocateGB could be refactored to a single function that takes size as a param

I have done this in my current new suballocator branch for the device mr_tests.cpp

harrism replied:
> `device_memory_resource` and `host_memory_resource` can't really shared the same tests because `host_memory_resource` allocation uses an `alignment` parameter whereas `device_memory_resource` alignment is fixed to 256. Likewise, the device memory resource tests test for if the memory is device memory, which will obviously fail for host memory resources.

Both things that could be factored out / specialized.

trevorsm7 commented on tests/mr/host/mr_tests.cpp:
```diff
@@ -0,0 +1,239 @@
+                });
+
+  std::for_each(allocations.begin(), allocations.end(),
+                [generator, distribution, this](allocation& a) {
```
```suggestion
                [this](allocation& a) {
```

vuule commented on include/rmm/mr/host/new_delete_resource.hpp:
```diff
@@ -0,0 +1,109 @@
+    }
+    char *ptr = static_cast<char *>(p);
+    // calculate where the offset is stored
+    std::size_t *offset = reinterpret_cast<std::size_t *>(ptr + bytes);
```
Trevor's suggestion applies here too:
```suggestion
    auto const offset = *reinterpret_cast<std::size_t *>(ptr + bytes);
```

vuule commented on include/rmm/mr/host/new_delete_resource.hpp:
```diff
@@ -0,0 +1,109 @@
+    // calculate where the offset is stored
+    std::size_t *offset = reinterpret_cast<std::size_t *>(ptr + bytes);
+    // calculate the original pointer
+    p = static_cast<void *>(ptr - *offset);
```
```suggestion
    p = static_cast<void *>(ptr - offset);
```

harrism replied:
I think you want the value stored at offset, not the pointer, no?

vuule commented on include/rmm/mr/host/pinned_memory_resource.hpp:
```diff
@@ -0,0 +1,110 @@
+    // calculate where the offset is stored
+    std::size_t *offset = reinterpret_cast<std::size_t *>(ptr + bytes);
+    // calculate the original pointer
+    p = static_cast<void *>(ptr - *offset);
```
```suggestion
    p = static_cast<void *>(ptr - offset);
```

jrhemstad commented on include/rmm/mr/host/new_delete_resource.hpp:
```diff
@@ -0,0 +1,109 @@
+    // calculate the return pointer
+    char *ptr = static_cast<char *>(p) + offset;
+    // store the offset right after the actually returned value
+    std::size_t *offset_store = reinterpret_cast<std::size_t *>(ptr + bytes);
```
Turns out this is actually undefined behavior. If `bytes % sizeof(size_t) != 0`, then this will result in a misaligned store (and misaligned load in the `do_deallocate`). 

Most CPU arches can handle misaligned load/stores (usually at a performance penalty), but according to the C++ standard it is UB. 

Instead, I'll store the offset immediately _prior_ to the returned pointer instead of at the end of the allocation.

harrism replied:
It's an arbitrary choice, but you could also align the location of the offset stored at the end of the allocation, right? Either one results in the same total storage?

OlivierNV replied:
Assuming this is intended for host code only, I think it's a reasonable assumption that the platform that interacts with a CUDA-capable GPU supports unaligned load/stores. This is a known, finite set of platforms.

jrhemstad replied:
> It's an arbitrary choice, but you could also align the location of the offset stored at the end of the allocation, right? Either one results in the same total storage?

I think it is arbitrary, yes, I was going to think through this tomorrow.

> Assuming this is intended for host code only, I think it's a reasonable assumption that the platform that interacts with a CUDA-capable GPU supports unaligned load/stores. This is a known, finite set of platforms.

Perhaps, but we should avoid UB when we can.

jrhemstad commented:
See refactored and improved aligned allocation functionality in `aligned.hpp` which provides `aligned_allocate` and `aligned_deallocate`.

This addresses most review comments as well as fixing the UB from the misaligned load/store.

jrhemstad commented:
rerun tests

trevorsm7 commented on include/rmm/mr/host/new_delete_resource.hpp:
```diff
@@ -42,29 +44,21 @@ class new_delete_resource final : public host_memory_resource {
-    return static_cast<void *>(ptr);
+
+    // If the requested alignment isn't supported, use default
+    alignment = (detail::is_supported_alignment(alignment))
```
Should there be an error or warning when the requested alignment isn't supported?

jrhemstad replied:
This comes straight from the documentation and requirements of `std::pmr::memory_resource`: 

> Allocates storage with a size of at least bytes bytes. The returned storage is aligned to the specified alignment if such alignment is supported, and to alignof(std::max_align_t) otherwise.

This is also documented in the `host_memory_resource`.

trevorsm7 replied:
Seems strange to fail silently, but I suppose that's fine if documented.

jrhemstad replied:
It's not failing. It's still allocating memory, but with a different alignment.

harrism commented on include/rmm/detail/aligned.hpp:
```diff
@@ -0,0 +1,127 @@
+  dealloc(original);
+}
+}  // namespace detail
+}  // namespace rmm
```
```suggestion
}  // namespace rmm

```
